families/track_and_trade/server/db/agents.js

```diff
@@ -0,0 +1,93 @@
+/**
+ * Copyright 2017 Intel Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ----------------------------------------------------------------------------
+ */
+'use strict'
+
+const r = require('rethinkdb')
+
+const db = require('./')
+
+const MAX_BLOCK = Number.MAX_SAFE_INTEGER
+
+const hasMaxBlock = obj => obj('endBlockNum').eq(MAX_BLOCK)
+
+const getAttribute = attr => obj => obj(attr)
+const getRecordId = getAttribute('recordId')
+const getPublicKey = getAttribute('publicKey')
+const getName = getAttribute('name')
+const getReporters = getAttribute('reporters')
+const getAuthorized = getAttribute('authorized')
+
+const hasPublicKey = key => obj => {
+  return r.eq(
+    key,
+    getPublicKey(obj)
+  )
+}
+
+const getAssociatedAgentId = role => record => record(role).nth(-1)('agentId')
+const getOwnerId = getAssociatedAgentId('owners')
+const getCustodianId = getAssociatedAgentId('custodians')
+
+const isAssociatedWithRecord = association => agent => record => {
+  return r.eq(
+    association(record),
+    getPublicKey(agent)
+  )
+}
+
+const isRecordOwner = isAssociatedWithRecord(getOwnerId)
+const isRecordCustodian = isAssociatedWithRecord(getCustodianId)
+
+const isReporter = agent => property => {
+  return getReporters(property)
+    .filter(hasPublicKey(getPublicKey(agent)))
+    .do(seq => r.branch(
+      seq.isEmpty(),
+      false,
+      getAuthorized(seq.nth(0))
+    ))
+}
+
+const getAgentsQuery = agents => {
+  return agents
+    .filter(hasMaxBlock)
```

Comment:
This has to query the block, and then query based on current block.  Otherwise, the query may result in agents that are in an inconsistent state from one block to the other.